vllm/v1/metrics/reader.py

```diff
@@ -0,0 +1,160 @@
+class Gauge(Metric):
+    """A numerical value that can go up or down."""
+    value: float
+
+
+@dataclass
+class Histogram(Metric):
+    """Observations recorded in configurable buckets.
+
+    Buckets are represented by a dictionary. The key is
+    the upper limit of the bucket, and the value is the
+    observed count in that bucket. A '+Inf' key always
+    exists.
+
+    The count property is the total count across all
+    buckets, identical to the count of the '+Inf' bucket.
+
+    The sum property is the total sum of all observed
+    values.
+    """
+    count: int
+    sum: float
+    buckets: dict[str, float]
+
+
+def get_metrics_snapshot() -> list[Metric]:
+    """An API for accessing in-memory Prometheus metrics.
+
+    Example:
+        >>> for metric in llm.get_metrics():
+        ...     if isinstance(metric, Counter):
+        ...         print(f"{metric} = {metric.value}")
+        ...     elif isinstance(metric, Gauge):
+        ...         print(f"{metric} = {metric.value}")
+        ...     elif isinstance(metric, Histogram):
+        ...         print(f"{metric}")
+        ...         print(f"    sum = {metric.sum}")
+        ...         print(f"    count = {metric.count}")
+        ...         for bucket_le, value in metrics.buckets.items():
+        ...             print(f"    {bucket_le} = {value}")
+    """
+    collected: list[Metric] = []
+    for metric in REGISTRY.collect():
+        if not metric.name.startswith("vllm:"):
+            continue
+        if metric.type == "gauge":
+            sample = _must_get_sample(metric)
+            collected.append(
+                Gauge(name=metric.name,
+                      labels=sample.labels,
+                      value=sample.value))
+        elif metric.type == "counter":
+            samples = _get_samples(metric, "_total")
+            if metric.name == "vllm:spec_decode_num_accepted_tokens_per_pos":
+                #
+                # Ugly vllm:num_accepted_tokens_per_pos special case.
+                #
+                # This metric is a vector of counters - for each spec
+                # decoding token position, we observe the number of
+                # accepted tokens using a Counter labeled with 'position'.
+                # We convert these into a vector of integer values.
+                #
+                assert metric.name == \
+                    "vllm:spec_decode_num_accepted_tokens_per_pos"
+                values: list[int] = [0] * len(samples)
+                for s in samples:
+                    values[int(s.labels["position"])] = int(s.value)
+                collected.append(
+                    Vector(name=metric.name,
+                           labels=_strip_label(s.labels, "position"),
+                           values=values))
+                continue
+
+            for s in samples:
+                collected.append(
+                    Counter(name=metric.name,
+                            labels=samples[0].labels,
```

Comment:
QQ: Can we safely get the first sample here?

Reply:
Great question - basically a bug that you'd only see with a DP config (multiple samples labelled with different `engine_id`)